### ISSUE TITLE
[API Proposal]: Parallel Foreach

### ISSUE THREAD

**Issue by Mr0N** [dotnet/runtime#99864]
### Background and motivation

I suggest adding a new class in .NET or replacing Parallel.ForEach. The main idea behind replacing the class lies in the fact that the implementation of Parallel.ForEach improperly utilizes resources. It takes a collection and inserts each element into the thread pool one by one, thereby affecting the efficiency of its operation due to the excessive creation of Task class instances. This is the first, albeit not critical, problem with the implementation of Foreach. The second problem is that resources and the code itself jump between different threads.

In other words, if you create an instance of the HttpClient class and insert it into the middle of Parallel.ForEach, it will be passed between different threads. I took HttpClient as an example; it is thread-safe, but there are other classes that are not thread-safe. And to make them work within Parallel.ForEach, you need to use a lock to block the thread.

In my implementation, the collection is divided into parts and distributed among different threads, ensuring that resources are not scattered across different threads and do not need to be synchronized. Plus, there's no need to synchronize resources for different threads.

### API Proposal

```csharp
namespace ParallelLib
{
    public class Parallel
    {
        private (CancellationToken token, CancellationTokenSource factory) CreateToken()
        {
            var factoryToken = new CancellationTokenSource();
            return (factoryToken.Token, factoryToken);
        }
        public void Foreach<TObj, T>(IEnumerable<TObj> enumerable,
            Action<TObj?, T> func,
            Func<T> createDependencies,
            int countTask
            ) where T : IDisposable
        {

            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        using var dep = createDependencies();
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item, dep);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        public void Foreach<TObj, R>(IEnumerable<TObj> enumerable,
           Action<TObj?, R> func,
           int countTask,
           Func<R> createDependencies)
        {
            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        var dep = createDependencies();
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item, dep);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        public void Foreach<TObj>(IEnumerable<TObj> enumerable,
            Action<TObj?> func, int countTask)
        {
            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        private IEnumerable<IEnumerable<TObj>> Slice<TObj>(IEnumerable<TObj> en, int count)
        {
            var enumerator = en.GetEnumerator();
            for (int i = 0; i < count; i++)
            {
                yield return GetPart(enumerator);
            }

        }
        private IEnumerable<TObj> GetPart<TObj>(IEnumerator<TObj> en)
        {
            while (true)
            {
                lock (objLock)
                {
                    if (!en.MoveNext())
                        break;
                }
                yield return en.Current;
            }
        }
        object objLock = new object();
    }
}

```

### API Usage

```csharp
  var paraller = new Parallel();
  paraller.Foreach(Enumerable.Range(0, 1000), a =>
  {
      throw new Exception();
  }, 10);
```

**Comment by Clockwork-Muse** [comment]
We can't remove it, that would be a breaking change.

>  It takes a collection and inserts each element into the thread pool one by one, thereby affecting the efficiency of its operation due to the excessive creation of Task class instances.

This isn't true.

> In my implementation...

You can already do something like this [with an existing overload](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreach?view=net-8.0#system-threading-tasks-parallel-foreach-2(system-collections-generic-ienumerable((-0))-system-threading-tasks-paralleloptions-system-func((-1))-system-func((-0-system-threading-tasks-parallelloopstate-1-1))-system-action((-1)))).

**Comment by Mr0N** [comment]
> This isn't true.
> 
> >

What exactly is not true?

**Comment by Mr0N** [comment]
> We can't remove it, that would be a breaking change.
> 
> >

You can either create another class for parallel foreach, or add another foreach method to Linq.

**Comment by svick** [comment]
> What exactly is not true?

`Parallel.ForEach` does not process items one-by-one, each in its own `Task` (at least not by default). Instead, it uses a `Partitioner` to efficiently split the collection between tasks.

You can see that with code like:

```c#
var taskCounts = new ConcurrentDictionary<int, int>();

Parallel.ForEach(Enumerable.Range(0, 1000), _ =>
{
    var id = Task.CurrentId.Value;
    taskCounts.AddOrUpdate(id, _ => 1, (_, n) => n + 1);
});

taskCounts.Dump();
```

The results on my machine are variable, but there is usually just a handful of `Task`s handling all the thousand items.

**Comment by Mr0N** [comment]
> > What exactly is not true?
> 
> `Parallel.ForEach` does not process items one-by-one, each in its own `Task` (at least not by default). Instead, it uses a `Partitioner` to efficiently split the collection between tasks.
> 
> You can see that with code like:
> 
> ```cs
> var taskCounts = new ConcurrentDictionary<int, int>();
> 
> Parallel.ForEach(Enumerable.Range(0, 1000), _ =>
> {
>     var id = Task.CurrentId.Value;
>     taskCounts.AddOrUpdate(id, _ => 1, (_, n) => n + 1);
> });
> 
> taskCounts.Dump();
> ```
> 
> The results on my machine are variable, but there is usually just a handful of `Task`s handling all the thousand items.

```
using System.Collections.Concurrent;

var taskCounts = new ConcurrentDictionary<Info, int>();

Parallel.ForEach(Enumerable.Range(0, 100), _ =>
{
    Thread.Sleep(1000);
    var id = Task.CurrentId.Value;
    taskCounts.AddOrUpdate(new Info(Thread.CurrentThread.ManagedThreadId,id), _ => 1, (_, n) => n + 1);
});
foreach (var item in taskCounts)
{
    Console.WriteLine($"Task:{item.Key} Elements:{item.Key}");
}
Console.ReadKey();


public record Info(int ThreadId,int TaskId);
![image](https://github.com/dotnet/runtime/assets/61237140/a8d4e49e-70d1-4492-9275-7507836175dd)

```

**Comment by Mr0N** [comment]
![image](https://github.com/dotnet/runtime/assets/61237140/f5c78af1-e3f8-4e9f-a29e-0723fbec6006)
The job is very strange; it throws data through various streams. Well, it seems that a separate task is not created for each iteration.

**Comment by Mr0N** [comment]
In my implementation example of Foreach, there is a specified number of threads in which Foreach works, thus it doesn't duplicate resources.

**Comment by Mr0N** [comment]
![image](https://github.com/dotnet/runtime/assets/61237140/f49664c4-bae0-40e2-9dfd-95b046629e75)

**Comment by svick** [comment]
@Mr0N If you spend a second blocking the thread, then the overhead of creating a `Task` is negligible, and `Parallel.ForEach` correctly decided that processing each item separately is the most efficient approach.

With limited `MaxDegreeOfParallelism`, there are tens of unnecessary allocations, which is something that could be improved. But that's again negligible on this scale and certainly wouldn't require a new API.

**Comment by Mr0N** [comment]
> @Mr0N If you spend a second blocking the thread, then the overhead of creating a `Task` is negligible, and `Parallel.ForEach` correctly decided that processing each item separately is the most efficient approach.

"Parallel.ForEach" doesn't create as many tasks and threads, if instead of "Thread.Sleep" there is a "for" loop, the computer may simply freeze.




> With limited `MaxDegreeOfParallelism`, there are tens of unnecessary allocations, which is something that could be improved. But that's again negligible on this scale and certainly wouldn't require a new API.

I don't like the concept where the body of Parallel.ForEach executes concurrently and it's unclear in which thread it's running. I would prefer definitiveness and stability. In the example code above, everything is executed within a single task, resources are not moved between different threads, and synchronization is not needed. Currently, when you want to put something inside the body of Parallel.ForEach, you have to think about how it will work under such instability.

Essentially, the code above is a regular foreach loop, as in C#, where you don't even have to think about the thread safety of the collection itself and its implementation inside

**Comment by svick** [comment]
> In the example code above, everything is executed within a single task, resources are not moved between different threads, and synchronization is not needed.

That's what the overloads with `TLocal` are for, as mentioned above.

> Currently, when you want to put something inside the body of Parallel.ForEach, you have to think about how it will work under such instability.

When you're writing parallel code, you always have to think about thread-safety, there is no avoiding that.

>  where you don't even have to think about the thread safety of the collection itself

You don't, that's handled by the `Paritioner`.

**Comment by Mr0N** [comment]
> That's what the overloads with `TLocal` are for, as mentioned above.
> 
> >

The TLocal seems like some sort of workaround.

And on the screenshot, it's evident that Parallel.ForEach created 27 resources for 27 tasks. However, there are significantly fewer threads. This means it's not utilizing resources efficiently.
```

using System.Collections.Concurrent;

var taskCounts = new ConcurrentDictionary<Info, int>();
Parallel.ForEach(Enumerable.Range(0, 100),
    new ParallelOptions(),
    () => new TestObj(),
    (q, t, d) =>
{
    Thread.Sleep(1000);
    var id = Task.CurrentId.Value;
    taskCounts.AddOrUpdate(new Info(Thread.CurrentThread.ManagedThreadId, id), _ => 1, (_, n) => n + 1);
    return d;
}, a => Console.WriteLine(a));
foreach (var item in taskCounts)
{
    Console.WriteLine($"Task:{item.Key} Elements:{item.Value}");
}
Console.WriteLine(TestObj.Count);
Console.ReadKey();

public class TestObj
{
    public static int Count { set; get; }
    public TestObj()
    {
        Count++;
    }
}


public record Info(int ThreadId, int TaskId);
```
![image](https://github.com/dotnet/runtime/assets/61237140/3211770f-85d5-4c7d-b67e-d2eb23020823)

**Comment by Mr0N** [comment]
> You don't, that's handled by the `Paritioner`.

I don't know exactly how the Partitioner processes the collection, but most likely it reads through the entire collection first, and then breaks it down into parts, into new sub-collections. In the code above, the collection is not read entirely and then divided into parts.

**Comment by xqrzd** [comment]
It looks like all you want to do is chunk the array, and process these chunks in parallel. Why not just chunk the input first?

```csharp
var chunks = Enumerable.Range(0, 100).Chunk(10);

Parallel.ForEach(chunks, collection =>
{
    foreach (var item in collection)
    {
        // Do something
    }
});
```

**Comment by Clockwork-Muse** [comment]
> The TLocal seems like some sort of workaround.

.... that's really the most reasonable way to handle it.  It's also what your proposed API does with the `createDependencies` parameter.

> And on the screenshot, it's evident that Parallel.ForEach created 27 resources for 27 tasks. However, there are significantly fewer threads. **This means it's not utilizing resources efficiently.**

This isn't true, at least not the way you think.  The task/thread system knows that the "thread" (Really task) stack has been put to sleep, so is running a different thread/task to more efficiently use CPU resources (and not necessarily whatever dependency resources each chunk requires).

It's too bad that there isn't an `async` version of the overload with `TLocal`....

> I don't know exactly how the Partitioner processes the collection, but most likely it reads through the entire collection first, and then breaks it down into parts, into new sub-collections. In the code above, the collection is not read entirely and then divided into parts.

Depends on the collection type.  For arrays it just uses the length of the array and splits it that way.  For some `IEnumerable` implementations it has to realize the array first, but it would have to do that anyways.

**Comment by Suchiman** [comment]
> I don't know exactly how the Partitioner processes the collection, but most likely it reads through the entire collection first, and then breaks it down into parts, into new sub-collections. In the code above, the collection is not read entirely and then divided into parts.

If you use the default overload of `Parallel.ForEach` that takes an IEnumerable, it will buffer it all up if it is not something IList like, and then partition it into ranges for each worker to process (see https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/custom-partitioners-for-plinq-and-tpl#kinds-of-partitioning ). You can also use one of the `Parallel.ForEach` overloads that takes a `Partitioner` instead of an `IEnumerable` to provide your own custom partitioning that does not buffer it all up to begin with.

**Comment by stephentoub** [comment]
> it will buffer it all up if it is not something IList like

This is definitely not the case. If it were, code like the following would deadlock and OOM, and that's obviously not what happens:
```C#
Parallel.ForEach(GetValues(), i =>
{
    Thread.Sleep(100);
    Console.WriteLine(i);
});

static IEnumerable<int> GetValues()
{
    for (int i = 0; ; i++) yield return i;
}
```

By default when you give Parallel.ForEach an `IEnumerable<T>`, it will check to see if it's a `T[]`/`IList<T>`. If it is, it will partition it based on ranges as was mentioned (though it doesn't just divide it up evenly by an expected number of workers... it's more dynamic than that). If you give it another kind of enumerable, it will itself use `Partitioner.Create(enumerable)` to create a partitioner. That default partitioner needs to use an enumerator to iterate through that enumerable, and because that's not thread-safe (at a minimum it needs to ensure the atomicity of two calls to MoveNext/Current), it has to lock around that access. Because that potentially has measurable overhead, it tries to ammortize the cost by buffering multiple items on each access. But, there's a flip side: grabbing multiple items ends up reserving those items for that consumer, which can lead to significant imbalance of work and lead to overall significant slowdowns (imagine if there were 16 workers, one worker reserved the next 16 work items, but it turns out there were only 16 work items total, and now one worker is tasked with doing all the work while the other 15 twiddle their thumbs, leaving the system drastically underutilized). As such, it employs a growth scheme, where each worker just takes a single item from the source, and then after a few times of doing that, each worker takes two items from the source, and then after a few times of doing that, each worker takes four items from the source, and so on, up to a maximum. This scheme makes it so that at first when we don't know how many items there will be, no one is greedy, but over time as it seems like more and more items will be available, workers become more liberal and are able to take more and more.  This, again, is the default behavior. If you don't like it, you can call Partitioner.Create yourself, specifying EnumerablePartitionerOptions.NoBuffering, and every worker will only take a single item on each call: that maximizes the chances for work being load-balanced well, but increases the per-item overhead by requiring a lock be taken and released on each access.

> It takes a collection and inserts each element into the thread pool one by one, thereby affecting the efficiency of its operation due to the excessive creation of Task class instances

As was mentioned, this is also definitely not the case. Parallel.ForEach works by creating a single Task, and when that Task is executed, it both creates a replica of itself and then starts processing the loop. If/when that replica is invoked, it will do the same thing: create a replica of itself, and then join in the processing of the loop. If/when that replica is invoked, same thing: it will create a replica of itself, and then join in the processing of the loop (this is all subject to the max degree of parallelism, such that a new replica won't be created if the limit has been reached). In this manner, ForEach is able to spread itself across any threads able to pick up a pending work item. It most certainly does not proactively just create a new Task per item in the source. But there's another wrinkle. Parallel.ForEach wants to be a good thread pool citizen, and the thread pool wants to be able to quickly and easily add and remove threads from the rotation; this is how it enables hill climbing, constantly running experiments like "if I add another thread or if I remove a thread from servicing work, does that increase or decrease work item completion throughput". If ForEach commandeered a thread and held onto it for the duration of a very long execution, while also commandeering every other thread that joins in with its replicas, the thread pool's mechanism would fail. As such, the Tasks ForEach creates are gracious and give the pool an opportunity to reclaim the thread. Every now and then, rather than just blindly processing the next item from the input, a Task will queue a replica of itself and then exit. That releases the thread back to the thread pool's loop, which allows the pool to choose to do something else with that thread.  That's likely why you see a few extra Tasks created.

> In my implementation example of Foreach, there is a specified number of threads in which Foreach works, thus it doesn't duplicate resources.

If you really want to use Parallel.ForEach with a fixed number of work items that never go away, you can, e.g.
```C#
Parallel.ForEach(Partitioner.Create(source).GetPartitions(N), p =>
{
    using var _ = p;
    while (p.MoveNext())
        body(p.Current);
});
```
That will create N partitions over the source, with ForEach running up to N loops handling the partitions. If you want additional state for each loop, have at it. But we don't have any plans to add such a thing to the core libraries. As always, if you believe your implementation is something others would benefit from, I'd encourage you to publish it as a nuget package so that others can use it.

**Comment by Suchiman** [comment]
> This is definitely not the case. If it were, code like the following would deadlock and OOM

I vaguely remember that, in the past, i had to use `Partitioner.Create(EnumerablePartitionerOptions.NoBuffering)` because i had very similiar code to what you are showing (with the yield) and the issue was, that the processing of the ForEach body would not start for a long time, maybe the chunk size was just too big which i've wrongly remembered as it "buffering entirely".

**Comment by Mr0N** [comment]
> By default when you give Parallel.ForEach an `IEnumerable<T>`, it will check to see if it's a `T[]`/`IList<T>`. If it is, it will partition it based on ranges as was mentioned (though it doesn't just divide it up evenly by an expected number of workers... it's more dynamic than that). If you give it another kind of enumerable, it will itself use `Partitioner.Create(enumerable)` to create a partitioner. That default partitioner needs to use an enumerator to iterate through that enumerable, and because that's not thread-safe (at a minimum it needs to ensure the atomicity of two calls to MoveNext/Current), it has to lock around that access. Because that potentially has measurable overhead, it tries to ammortize the cost by buffering multiple items on each access. But, there's a flip side: grabbing multiple items ends up reserving those items for that consumer, which can lead to significant imbalance of work and lead to overall significant slowdowns (imagine if there were 16 workers, one worker reserved the next 16 work items, but it turns out there were only 16 work items total, and now one worker is tasked with doing all the work while the other 15 twiddle their thumbs, leaving the system drastically underutilized). As such, it employs a growth scheme, where each worker just takes a single item from the source, and then after a few times of doing that, each worker takes two items from the source, and then after a few times of doing that, each worker takes four items from the source, and so on, up to a maximum. This scheme makes it so that at first when we don't know how many items there will be, no one is greedy, but over time as it seems like more and more items will be available, workers become more liberal and are able to take more and more. This, again, is the default behavior. If you don't like it, you can call Partitioner.Create yourself, specifying EnumerablePartitionerOptions.NoBuffering, and every worker will only take a single item on each call: that maximizes the chances for work being load-balanced well, but increases the per-item overhead by requiring a lock be taken and released on each access.

Here is one problem, as if everything is not deterministic, the class behaves differently depending on the input data. That is, the user cannot understand how the class will work with these input data.







> As was mentioned, this is also definitely not the case. Parallel.ForEach works by creating a single Task, and when that Task is executed, it both creates a replica of itself and then starts processing the loop. If/when that replica is invoked, it will do the same thing: create a replica of itself, and then join in the processing of the loop. If/when that replica is invoked, same thing: it will create a replica of itself, and then join in the processing of the loop (this is all subject to the max degree of parallelism, such that a new replica won't be created if the limit has been reached). In this manner, ForEach is able to spread itself across any threads able to pick up a pending work item. It most certainly does not proactively just create a new Task per item in the source. But there's another wrinkle. Parallel.ForEach wants to be a good thread pool citizen, and the thread pool wants to be able to quickly and easily add and remove threads from the rotation; this is how it enables hill climbing, constantly running experiments like "if I add another thread or if I remove a thread from servicing work, does that increase or decrease work item completion throughput". If ForEach commandeered a thread and held onto it for the duration of a very long execution, while also commandeering every other thread that joins in with its replicas, the thread pool's mechanism would fail. As such, the Tasks ForEach creates are gracious and give the pool an opportunity to reclaim the thread. Every now and then, rather than just blindly processing the next item from the input, a Task will queue a replica of itself and then exit. That releases the thread back to the thread pool's loop, which allows the pool to choose to do something else with that thread. That's likely why you see a few extra Tasks created.



I don't like it, the class wants to load the entire CPU, given that it does not know what is being executed in the body of the foreach loop. A regular foreach loop doesn't think about anything, it just performs enumeration of the collection.







> ```cs
> Parallel.ForEach(Partitioner.Create(source).GetPartitions(N), p =>
> {
>     using var _ = p;
>     while (p.MoveNext())
>         body(p.Current);
> });
> ```

Well, this doesn't look like a foreach loop anymore.

**Comment by Mr0N** [comment]
> Reference

And here, for example, there's a bunch of resource creation going on. Some resources are simply created, while others can be created in limited quantities. Under such conditions, one has to create various factories to somehow distribute them there, which makes it easier not to use Parallel.ForEach.

**Comment by Mr0N** [comment]
> > That's what the overloads with `TLocal` are for, as mentioned above.
> > >
> 
> The TLocal seems like some sort of workaround.
> 
> And on the screenshot, it's evident that Parallel.ForEach created 27 resources for 27 tasks. However, there are significantly fewer threads. This means it's not utilizing resources efficiently.
> 
> ```
> 
> using System.Collections.Concurrent;
> 
> var taskCounts = new ConcurrentDictionary<Info, int>();
> Parallel.ForEach(Enumerable.Range(0, 100),
>     new ParallelOptions(),
>     () => new TestObj(),
>     (q, t, d) =>
> {
>     Thread.Sleep(1000);
>     var id = Task.CurrentId.Value;
>     taskCounts.AddOrUpdate(new Info(Thread.CurrentThread.ManagedThreadId, id), _ => 1, (_, n) => n + 1);
>     return d;
> }, a => Console.WriteLine(a));
> foreach (var item in taskCounts)
> {
>     Console.WriteLine($"Task:{item.Key} Elements:{item.Value}");
> }
> Console.WriteLine(TestObj.Count);
> Console.ReadKey();
> 
> public class TestObj
> {
>     public static int Count { set; get; }
>     public TestObj()
>     {
>         Count++;
>     }
> }
> 
> 
> public record Info(int ThreadId, int TaskId);
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/61237140/313496641-3211770f-85d5-4c7d-b67e-d2eb23020823.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA3NTcwOTksIm5iZiI6MTcxMDc1Njc5OSwicGF0aCI6Ii82MTIzNzE0MC8zMTM0OTY2NDEtMzIxMTc3MGYtODVkNS00YzdkLWI2N2UtZDJlYjIzMDIwODIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE4VDEwMTMxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkODExYWFiNjBlN2FjYjAzN2NhYjlhNjcwMDAxZTgzMjQ1Y2FiZjk3MzU4Yjk2M2I2ZmRmMTI4YWFlNTE4NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0dvI3x2vGl5L5TQRBzNCSFIs6WOb6Xbt_X9S9GnJQsg)

And here, for example, there's a bunch of resource creation going on. Some resources are simply created, while others can be created in limited quantities. Under such conditions, one has to create various factories to somehow distribute them there, which makes it easier not to use Parallel.ForEach.

**Comment by stephentoub** [comment]
> I don't like it,

You're free to hide it in a 10-line helper in your own projects or your own nuget package.

**Comment by Mr0N** [comment]
In terms of using Parallel.ForEach, it's too complicated to use.

1)The default reload puts 100 percent load on the processor.
2)Resource creation is too complex, and if their quantity is limited, it becomes a very difficult problem.
3)It's unclear what to expect from it during operation; it creates a very large number of tasks that use the same threads as before. In my implementation, five threads are taken that work until the foreach operation is complete.

**Comment by stephentoub** [comment]
> The default reload puts 100 percent load on the processor.

By design. If you want to use less, you specify an explicit MaxDegreeOfParallelism.

> In my implementation, five threads are taken that work until the foreach operation is complete

Which, as I noted in https://github.com/dotnet/runtime/issues/99864#issuecomment-2002693454, we very explicitly try hard to have that _not_ be the case, in order to allow the thread pool to better manage its threads and workload. Any other such addition to the core libraries would do something similar as well.

> it's too complex to use

If it's not your cup of tea, you're welcome of course to not use it. As I noted, if you have your own helper you prefer, you can keep using it, and if you believe others would like it, wrap it up as a nuget package and publish it (though before you do, you'll want to fix issues like the access to Current in the original example not being part of the lock block).

I appreciate your interest and suggestion here, but we're not going to add another ParallelForEach as proposed. Thanks.

**Comment by Mr0N** [comment]
> > The default reload puts 100 percent load on the processor.
> 
> By design. If you want to use less, you specify an explicit MaxDegreeOfParallelism.
> 
> > In my implementation, five threads are taken that work until the foreach operation is complete
> 
> Which, as I noted in [#99864 (comment)](https://github.com/dotnet/runtime/issues/99864#issuecomment-2002693454), we very explicitly try hard to have that _not_ be the case, in order to allow the thread pool to better manage its threads and workload. Any other such addition to the core libraries would do something similar as well.
> 
> > it's too complex to use
> 
> If it's not your cup of tea, you're welcome of course to not use it. As I noted, if you have your own helper you prefer, you can keep using it, and if you believe others would like it, wrap it up as a nuget package and publish it (though before you do, you'll want to fix issues like the access to Current in the original example not being part of the lock block).
> 
> I appreciate your interest and suggestion here, but we're not going to add another ParallelForEach as proposed. Thanks.

"Current" is a property where there is no need for any logic regarding the acquisition or processing of information; it doesn't make sense to block it.

**Comment by Mr0N** [comment]
Well, okay, in theory, it's possible for two threads to take the same information. But it's unlikely, and we need to test whether it's possible.

**Comment by stephentoub** [comment]
> But it's unlikely, and we need to test whether it's possible.

It's absolutely possible and, under contention, likely.

**Comment by stephentoub** [comment]
Here's a simple example:
```C#
const int Iters = 10_000_000;

IEnumerable<int> source = Enumerable.Range(0, Iters);
using IEnumerator<int> e = source.GetEnumerator();

HashSet<int> ints1 = new(), ints2 = new();
await Task.WhenAll(from ints in new[] { ints1, ints2 } select Task.Run(() =>
{
    while (true)
    {
        lock (e)
        {
            if (!e.MoveNext()) break;
        }
        ints.Add(e.Current); // bug
    }
}));

ints1.UnionWith(ints2);

Console.WriteLine($"Expected: {Iters}, Actual: {ints1.Count}");
```
That frequently prints output like:
```
Expected: 10000000, Actual: 9998939
```
when those numbers should be the same if the program were correct.

**Comment by Mr0N** [comment]
> Here's a simple example:
> 
> ```cs
> const int Iters = 10_000_000;
> 
> IEnumerable<int> source = Enumerable.Range(0, Iters);
> using IEnumerator<int> e = source.GetEnumerator();
> 
> HashSet<int> ints1 = new(), ints2 = new();
> await Task.WhenAll(from ints in new[] { ints1, ints2 } select Task.Run(() =>
> {
>     while (true)
>     {
>         lock (e)
>         {
>             if (!e.MoveNext()) break;
>         }
>         ints.Add(e.Current); // bug
>     }
> }));
> 
> ints1.UnionWith(ints2);
> 
> Console.WriteLine($"Expected: {Iters}, Actual: {ints1.Count}");
> ```
> 
> That frequently prints output like:
> 
> ```
> Expected: 10000000, Actual: 9998939
> ```
> 
> when those numbers should be the same if the program were correct.
The yield won't allow the thread to behave like that when it enters the method; the method will be in some state, either for fetching data or for MoveNext.
![image](https://github.com/dotnet/runtime/assets/61237140/8e20777b-4be8-4d28-bef0-1c933bd5cc90)







https://sharplab.io/#gist:1003a29fe1bd422a0d8b39d92eb9bf9e

**Comment by Mr0N** [comment]
Well, in principle, I'm not really liking this separation anymore. I'll come up with something else.

**Comment by Mr0N** [comment]
Yes, there is a bug; if you block the Current, everything is fine. So, I wrote an issue so that something similar would be in .NET, tested, and so on...






```
object objLock = new object();
var en = Enumerable.Range(1, 10_000_000);
ConcurrentBag<int> ls = new ConcurrentBag<int>();
var tasks = new List<Task>();

var res = Slice(en, 10);
foreach (var item in res)
{
    var task = Task.Run(() =>
    {
        Thread.Sleep(500);
        foreach (var element in item)
        {
            ls.Add(element);
        }
    });
    tasks.Add(task);
}
Task.WaitAll(tasks.ToArray());
int number = 0;
foreach (var item in ls.OrderBy(a => a))
{
    number++;
    if (item != number)
    {
        throw new Exception("Exception");
    }
}
Console.WriteLine($"Number:{number} All Count:{ls.Count}");
Console.ReadKey();
IEnumerable<IEnumerable<TObj>> Slice<TObj>(IEnumerable<TObj> en, int count)
{
    var enumerator = en.GetEnumerator();
    for (int i = 0; i < count; i++)
    {
        yield return GetPart(enumerator);
    }

}
IEnumerable<TObj> GetPart<TObj>(IEnumerator<TObj> en)
{
    while (true)
    {
        lock (objLock)
        {
            if (!en.MoveNext())//
                break;
        }
        yield return en.Current;
       
    }
}

```

**Comment by Mr0N** [comment]
working version
```
    public class Parallel
    {
        private (CancellationToken token, CancellationTokenSource factory) CreateToken()
        {
            var factoryToken = new CancellationTokenSource();
            return (factoryToken.Token, factoryToken);
        }
        public void Foreach<TObj, T>(IEnumerable<TObj> enumerable,
            Action<TObj?, T> func,
            Func<T> createDependencies,
            int countTask
            ) where T : IDisposable
        {

            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        using var dep = createDependencies();
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item, dep);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        public void Foreach<TObj, R>(IEnumerable<TObj> enumerable,
           Action<TObj?, R> func,
           int countTask,
           Func<R> createDependencies)
        {
            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        var dep = createDependencies();
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item, dep);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        public void Foreach<TObj>(IEnumerable<TObj> enumerable,
            Action<TObj?> func, int countTask)
        {
            var query = Slice(enumerable, countTask);
            var ls = new List<Task>(countTask);
            var token = CreateToken();
            foreach (var iCollection in query)
            {
                var task = Task.Run(() =>
                {
                    try
                    {
                        foreach (var item in iCollection)
                        {
                            if (token.token.IsCancellationRequested) break;
                            func.Invoke(item);
                        }
                    }
                    catch (Exception ex)
                    {
                        token.factory.Cancel();
                        throw new Exception("", ex);
                    }
                });
                ls.Add(task);
            }
            Task.WaitAll(ls.ToArray());
            // enumerable.Chunk
        }
        private IEnumerable<IEnumerable<TObj>> Slice<TObj>(IEnumerable<TObj> en, int count)
        {
            var enumerator = en.GetEnumerator();
            for (int i = 0; i < count; i++)
            {
                yield return GetPart(enumerator);
            }

        }
       private IEnumerable<TObj> GetPart<TObj>(IEnumerator<TObj> en)
        {
            TObj obj = default;
            while (true)
            {
                lock (objLock)
                {
                    if (!en.MoveNext())//
                        break;
                    obj = en.Current;
                }
                //lock (objLock)

                yield return obj;


            }
        }
        object objLock = new object();
    }
```

**Comment by Mr0N** [comment]
This works, checked with reference and structural types.
```
object objLock = new object();
var en = Enumerable.Range(1, 10_000_000);
ConcurrentBag<string> ls = new ConcurrentBag<string>();
var tasks = new List<Task>();

var res = Slice(en, 10);
foreach (var item in res)
{
    var task = Task.Run(() =>
    {
        Thread.Sleep(500);
        foreach (var element in item)
        {
            ls.Add(element.ToString());
        }
    });
    tasks.Add(task);
}
Task.WaitAll(tasks.ToArray());
int number = 0;
var order = ls.OrderBy(a => int.Parse(a)).ToList();
foreach (var item in order)
{
    number++;
    if (item != number.ToString())
    {
        throw new Exception("Exception");
    }
}
Console.WriteLine($"Number:{number} All Count:{ls.Count}");
Console.ReadKey();
IEnumerable<IEnumerable<TObj>> Slice<TObj>(IEnumerable<TObj> en, int count)
{
    var enumerator = en.GetEnumerator();
    for (int i = 0; i < count; i++)
    {
        yield return GetPart(enumerator);
    }

}

IEnumerable<TObj> GetPart<TObj>(IEnumerator<TObj> en)
{
    TObj obj = default;
    while (true)
    {
        lock (objLock)
        {
            if (!en.MoveNext())//
                break;
            obj = en.Current;
        }
        yield return obj;
    }
   
}


```
